lib/core/src/Cardano/Pool/Metrics.hs

```
@@ -52,8 +73,105 @@ import GHC.Generics
+
+    follow nl tr knownHeaders advance' rollback getHeader
+  where
+
+    advance'
+        :: NE.NonEmpty Block
+        -> BlockHeader
+        -> IO (FollowAction ErrMetricsWorkerErr)
+    advance' bs h = runExceptT (advance bs h) >>= return . \case
+        Left ErrMetricsWorkerErrTipTooVolatile
+            -> Retry
+        Left e
+            -> ExitWith e
+            -- TODO: Make all cases explicit here.
+        Right ()
+            -> Continue
+
+    advance
+        :: NE.NonEmpty Block
+        -> BlockHeader
+        -> ExceptT ErrMetricsWorkerErr IO ()
+    advance blocks nodeTip = do
+        -- Get a stake distribution and next batch of blocks such that we know
+        -- they are from the same chain.
+        --
+        -- Assumtion: We don't expect the node to switch from tip A to tip B,
+        -- and back to tip A.
+        --
+        -- If the tip is still the same after we have fetched the stake
+        -- distribution, we conclude that the stake-distrubtion is from the same
+        -- chain as the tip.
+        consistentDistr <- withinSameTip retryPolicy liftE getTip $ \nodeTip2 -> do
+            distr <- withExceptT (const ErrMetricsWorkerStakeDistributionUnavailible)
+                $ stakeDistribution nl
+            unless (nodeTip == nodeTip2) $
+                -- The tip has changed. We cannot guarantee that the blocks and
+                -- the stake-distribution are on the same chain.
+                throwE ErrMetricsWorkerErrTipTooVolatile
+            return (fst distr, Map.toList $ snd distr)
+
+        liftIO $ uncurry (putStakeDistribution db) consistentDistr
+
+        forM_ blocks $ \block ->
+            -- TODO: We might want to batch the insertion to the DB.
+            case producer block of
+                Just pool ->
+                    -- TODO: What do we do if the production already exists?!
+                    withExceptT (const ErrMetricsWorkerDataInconsistencyDuplicateBlock) $
```

We can't recover from that (it means that either our roll back mechanism or chain follower is broken). This is a 500 and a valid case for Exception.

---

Note that the `worker` (collecting metrics and writing to DB) will now be completely separate from a future `StakePoolLayer` (reading from DB).

I guess we should recover by clearing the `pool_production` table in the DB.

---

Hmmm. No. This will silently swallow a data inconsistency issue that shouldn't have happened. If this happens, we must crash.

---

>  we must crash

Crash what exactly? The worker thread? The entire wallet? (we would need added logic for the latter I believe)

---

We could also return `ExitOn ErrMetricsWorkerDataInconsistencyDuplicateBlock` to the callback such that it stops.